Wikipedia/Code/SavedArticlesFetcher.m

```diff
@@ -170,11 +172,94 @@ - (void)downloadImageDataForArticle:(MWKArticle*)article failure:(WMFErrorHandle
+        
+        //if this is the article image, skip over it because it should be preserved
+        BOOL isArticleImage = [article.imageURL isEqualToString:imageURLString];
+        if (isArticleImage) {
+            continue;
+        }
+        
+        NSUInteger width = WMFParseSizePrefixFromSourceURL(imageURLString);
+        if (width != articleImageWidth) {
+            NSURL *imageURL = [NSURL URLWithString:imageURLString];
+            
+            if ([imageController hasDataOnDiskForImageWithURL:imageURL]) {
+                NSString *cachedPath = [imageController cachePathForImageWithURL:imageURL];
+                
+                NSString *articleURLString = WMFChangeImageSourceURLSizePrefix(imageURLString, articleImageWidth);
+                NSURL *articleURL = [NSURL URLWithString:articleURLString];
+                if (![imageController hasDataOnDiskForImageWithURL:articleURL]) {
+                    NSString *articlePath = [imageController cachePathForImageWithURL:articleURL];
+                    NSError *copyError = nil;
+                    if (![fileManager copyItemAtPath:cachedPath toPath:articlePath error:&copyError]) {
+                        DDLogError(@"Error copying cached image to article image: %@", copyError);
+                    }
+                }
+                
+                NSString *originalURLString = WMFOriginalImageURLStringFromURLString(imageURLString);
+                NSURL *originalURL = [NSURL URLWithString:originalURLString];
+                if (![imageController hasDataOnDiskForImageWithURL:originalURL]) {
+                    NSString *originalPath = [imageController cachePathForImageWithURL:originalURL];
+                    NSError *copyError = nil;
+                    if (![fileManager copyItemAtPath:cachedPath toPath:originalPath error:&copyError]) {
+                        DDLogError(@"Error copying cached image to original image: %@", copyError);
+                    }
+                    
+                }
+                
+                if (width != NSNotFound) { //remove the cached image only if this isn't the original file
+                    NSError *removalError = nil;
+                    if (![fileManager removeItemAtPath:cachedPath error:&removalError]) {
+                        DDLogError(@"Error removing legacy cached image: %@", removalError);
+                    }
+                }
+            }
+        }
+        
+        NSError *removalError = nil;
+        if (![fileManager removeItemAtURL:imageFolderURL error:&removalError]) {
+            DDLogError(@"Error removing old image list image: %@", removalError);
+        }
+    }
+}
 
 - (void)fetchAllImagesInArticle:(MWKArticle*)article failure:(WMFErrorHandler)failure success:(WMFSuccessHandler)success {
+    WMF_TECH_DEBT_TODO(This legacy migration can be removed after enough users upgrade to 5.0.5)
+    [self migrateLegacyImagesInArticle:article];
```

Comment:
Or I guess you are doing it synchronously so it should be finished before any further logic is called - is that right?